agent/internal/taskoutput/test_log.go

```diff
@@ -0,0 +1,330 @@
+	}, testLog.Name, lines)
+}
+
+// TODO (DEVPROD-75): Remove this logic once we cut over to Evergreen logs.
+func sendTestLogToCedar(ctx context.Context, comm client.Communicator, tsk *task.Task, testLog *testlog.TestLog) error {
+	conn, err := comm.GetCedarGRPCConn(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "getting the Cedar gRPC connection for test '%s'", testLog.Name)
+	}
+
+	timberOpts := &buildlogger.LoggerOptions{
+		Project:    tsk.Project,
+		Version:    tsk.Version,
+		Variant:    tsk.BuildVariant,
+		TaskName:   tsk.DisplayName,
+		TaskID:     tsk.Id,
+		Execution:  int32(tsk.Execution),
+		TestName:   testLog.Name,
+		Mainline:   !tsk.IsPatchRequest(),
+		Storage:    buildlogger.LogStorageS3,
+		ClientConn: conn,
+	}
+	levelInfo := send.LevelInfo{Default: level.Info, Threshold: level.Debug}
+	sender, err := buildlogger.NewLoggerWithContext(ctx, testLog.Name, levelInfo, timberOpts)
+	if err != nil {
+		return errors.Wrapf(err, "creating buildlogger logger for test result '%s'", testLog.Name)
+	}
+
+	sender.Send(message.ConvertToComposer(level.Info, strings.Join(testLog.Lines, "\n")))
+	if err = sender.Close(); err != nil {
+		return errors.Wrapf(err, "closing buildlogger logger for test result '%s'", testLog.Name)
+	}
+
+	return nil
+}
+
+// testLogDirectoryHandler implements automatic and asynchronous task output
+// handling for the reserved test log directory.
+type testLogDirectoryHandler struct {
+	dir             string
+	logger          client.LoggerProducer
+	spec            testLogSpec
+	createSender    func(context.Context, string) (send.Sender, error)
+	cancel          context.CancelFunc
+	followFileCount int
+
+	once sync.Once
+	wg   sync.WaitGroup
+}
+
+// newTestLogDirectoryHandler returns a new test log directory handler for the
+// specified task.
+func newTestLogDirectoryHandler(output taskoutput.TestLogOutput, taskOpts taskoutput.TaskOptions, logger client.LoggerProducer) *testLogDirectoryHandler {
+	h := &testLogDirectoryHandler{logger: logger}
+	h.createSender = func(ctx context.Context, logPath string) (send.Sender, error) {
+		return output.NewSender(ctx, taskOpts, taskoutput.EvergreenSenderOptions{
+			Local: logger.Task().GetSender(),
+			Parse: h.spec.getParser(),
+		}, logPath)
+	}
+
+	return h
+}
+
+func (h *testLogDirectoryHandler) start(ctx context.Context, dir string) error {
+	h.dir = dir
+
+	watcherCtx, cancel := context.WithCancel(ctx)
+	h.cancel = cancel
+	go h.watch(watcherCtx)
+
+	return nil
+}
+
+// watch recursively watches the reserved test log directory for newly created
+// log files. When a new file is detected, an asynchronous file follower is
+// created to ingest the log data as it is written for the duration of the
+// task.
+func (h *testLogDirectoryHandler) watch(ctx context.Context) {
+	defer func() {
+		h.logger.Execution().Critical(recovery.HandlePanicWithError(recover(), nil, "test log directory watcher"))
+	}()
+
+	workerCtx, cancel := context.WithCancel(ctx)
```

Comment:
why do we need an internal context for watch?

Reply:
Ah that was from before, I will remove it.